### PR TITLE
Add Hkts, GenericComponents and GenericOperations

### PR DESCRIPTION
**Pull request by tjjfvi**
Waiting on #27

### CODE REVIEW

**Comment by tjjfvi** [comment]
Fixes #19

File: packages/core/src/Component.ts [hunk 1]
```diff
@@ -62,12 +62,34 @@ export const Component = {
...
+
+export const GenericComponent = {
+  create: <T, I extends Hkt<T, any[]>, O extends Hkt<T, Thing>>(
+    name: string,
+    func: <U extends T>(...args: Hkt.Output<I, U>) => Hkt.Output<O, U>,
+    opts?: ComponentOpts,
+  ): GenericComponent<T, I, O> =>
+    Component.create<Hkt.Output<I, T>, Hkt.Output<O, T>>(name, func, opts) as never,
+  applyHierarchy: <T, I extends Hkt<T, any[]>, O extends Hkt<T, Thing>>(
+    Component: GenericComponent<T, I, O>,
```
Comment (vjjft) [inline packages/core/src/Component.ts]: rename to `component`

File: packages/core/src/Component.ts [hunk 2]
```diff
@@ -62,12 +62,34 @@ export const Component = {
...
       hierarchy,
     }),
 }
+
+export type GenericComponent<T, I extends Hkt<T, any[]>, O extends Hkt<T, Thing>> =
+  & Omit<Component<Hkt.Output<I, T>, Hkt.Output<O, T>>, "func">
+  & {
+    readonly func: <U extends T>(...args: Hkt.Output<I, U>) => Hkt.Output<O, U>,
+    <U extends T>(...args: Hkt.Output<I, U>): StripRealm<Hkt.Output<O, U>>,
+    readonly __generic: true,
```
Comment (vjjft) [inline packages/core/src/Component.ts]: extract to unique symbol

File: packages/core/src/Operation.ts [hunk 3]
```diff
@@ -76,3 +77,25 @@ export const Operation = {
     }),
   isOperation: checkTypeProperty.string<Operation<any, any>>("Operation"),
 }
+
+export type GenericOperation<I extends Product, O extends Hkt<I, Product>> =
+  & Omit<Operation<I, Hkt.Output<O, I>>, "func">
+  & {
+    readonly func: <J extends I>(arg: Element<J>) => Elementish<Hkt.Output<O, J>>,
+    <J extends I>(...args: Elementish<J>[]): Element<Hkt.Output<O, I>>,
+    readonly __generic: undefined,
```
Comment (vjjft) [inline packages/core/src/Operation.ts]: extract to unique symbol

**Review by vjjft — DISMISSED** [review]

**Pull request merged**